Source/Core/VideoBackends/OGL/ProgramShaderCache.cpp

```diff
@@ -464,20 +535,52 @@ GLuint ProgramShaderCache::CompileSingleShader(GLuint type, const std::string& c
   {
     // Compile failed
     ERROR_LOG(VIDEO, "Shader compilation failed; see info log");
-
-    // Don't try to use this shader
-    glDeleteShader(result);
-    return 0;
+    return false;
   }
 
-  return result;
+  return true;
 }
 
-void ProgramShaderCache::GetShaderId(SHADERUID* uid, u32 primitive_type)
+bool ProgramShaderCache::CheckProgramLinkResult(GLuint id, const std::string& vcode,
+                                                const std::string& pcode, const std::string& gcode)
 {
-  uid->puid = GetPixelShaderUid();
-  uid->vuid = GetVertexShaderUid();
-  uid->guid = GetGeometryShaderUid(primitive_type);
+  GLint linkStatus;
+  glGetProgramiv(id, GL_LINK_STATUS, &linkStatus);
+  GLsizei length = 0;
+  glGetProgramiv(id, GL_INFO_LOG_LENGTH, &length);
+  if (linkStatus != GL_TRUE || (length > 1 && DEBUG_GLSL))
+  {
+    std::string info_log;
+    info_log.resize(length);
+    glGetProgramInfoLog(id, length, &length, &info_log[0]);
+    ERROR_LOG(VIDEO, "Program info log:\n%s", info_log.c_str());
+
+    std::string filename =
+        StringFromFormat("%sbad_p_%d.txt", File::GetUserPath(D_DUMP_IDX).c_str(), num_failures++);
+    std::ofstream file;
+    File::OpenFStream(file, filename, std::ios_base::out);
+    file << s_glsl_header << vcode << s_glsl_header << pcode;
+    if (!gcode.empty())
+      file << s_glsl_header << gcode;
+    file << info_log;
+    file.close();
+
+    if (linkStatus != GL_TRUE)
+    {
+      PanicAlert("Failed to link shaders: %s\n"
+                 "Debug info (%s, %s, %s):\n%s",
+                 filename.c_str(), g_ogl_config.gl_vendor, g_ogl_config.gl_renderer,
+                 g_ogl_config.gl_version, info_log.c_str());
+    }
+  }
+  if (linkStatus != GL_TRUE)
```

Comment:
Do we need this if check?  Do we typically keep a panic alert and an error log?  Seems like we should just return false above